File: DesktopClient/DesktopClient/ClientManager.cs
using System;
using System.Threading.Tasks;

namespace DesktopClient.Services {
    public class ClientManager {
        private const string CLIENT_TYPE = "desktop";

        public async Task<string> GetOrCreateClientIdAsync(string username, ApiClient apiClient) {
            try {
                // Sprawdź czy istnieje zapisane ID klienta dla tego użytkownika
                var savedClientId = GetSavedClientId(username);

                if (!string.IsNullOrEmpty(savedClientId)) {
                    // Sprawdź czy klient nadal istnieje na serwerze
                    if (await IsClientValidAsync(savedClientId, apiClient)) {
                        await apiClient.UpdateClientActivityAsync(savedClientId);
                        return savedClientId;
                    }

                    // Klient nie istnieje - usuń zapisane ID
                    RemoveSavedClientId(username);
                }

                // Zarejestruj nowego klienta
                var clientName = $"{Environment.MachineName} - {username}";
                var metadata = new {
                    machineName = Environment.MachineName,
                    userName = Environment.UserName,
                    osVersion = Environment.OSVersion.ToString()
                };

                var response = await apiClient.RegisterClientAsync(CLIENT_TYPE, clientName, metadata);

                if (response.success && response.client != null) {
                    SaveClientId(username, response.client.clientId);
                    return response.client.clientId;
                }

                throw new Exception("Nie udało się zarejestrować klienta");
            } catch (Exception ex) {
                throw new Exception($"Błąd podczas rejestracji klienta: {ex.Message}");
            }
        }

        private async Task<bool> IsClientValidAsync(string clientId, ApiClient apiClient) {
            try {
                var response = await apiClient.GetClientAsync(clientId);
                return response.success && response.client != null;
            } catch {
                return false;
            }
        }

        private string GetSavedClientId(string username) {
            try {
                var filePath = GetClientFilePath(username);
                if (System.IO.File.Exists(filePath)) {
                    return System.IO.File.ReadAllText(filePath).Trim();
                }
                return null;
            } catch {
                return null;
            }
        }

        private void SaveClientId(string username, string clientId) {
            try {
                var filePath = GetClientFilePath(username);
                var directory = System.IO.Path.GetDirectoryName(filePath);
                System.IO.Directory.CreateDirectory(directory);
                System.IO.File.WriteAllText(filePath, clientId);
            } catch {
                // Ignoruj błędy zapisu
            }
        }

        private void RemoveSavedClientId(string username) {
            try {
                var filePath = GetClientFilePath(username);
                if (System.IO.File.Exists(filePath)) {
                    System.IO.File.Delete(filePath);
                }
            } catch {
                // Ignoruj błędy usuwania
            }
        }

        private string GetClientFilePath(string username) {
            var appDataPath = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
            return System.IO.Path.Combine(appDataPath, "FileManager", $"{username}.client");
        }
    }
}

File: MobileClient/app/src/main/java/com/example/mobileclient/ClientManager.kt
// ClientManager.kt
package com.example.mobileclient

import android.content.Context
import android.os.Build
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.File

class ClientManager(private val context: Context) {

    private companion object {
        const val CLIENT_TYPE = "mobile"
        const val CLIENT_FILE_PREFIX = "client_"
    }

    suspend fun getOrCreateClientId(username: String, apiClient: ApiClient): String = withContext(Dispatchers.IO) {
        try {
            // Sprawdź czy istnieje zapisane ID klienta dla tego użytkownika
            val savedClientId = getSavedClientId(username)

            if (!savedClientId.isNullOrEmpty()) {
                // Sprawdź czy klient nadal istnieje na serwerze
                if (isClientValid(savedClientId, apiClient)) {
                    apiClient.updateClientActivity(savedClientId)
                    return@withContext savedClientId
                }

                // Klient nie istnieje - usuń zapisane ID
                removeSavedClientId(username)
            }

            // Zarejestruj nowego klienta
            val clientName = "${Build.MODEL} - $username"
            val metadata = mapOf(
                "deviceModel" to Build.MODEL,
                "deviceManufacturer" to Build.MANUFACTURER,
                "androidVersion" to Build.VERSION.RELEASE,
                "apiLevel" to Build.VERSION.SDK_INT,
                "deviceName" to (Build.DISPLAY ?: "Unknown")
            )

            val response = apiClient.registerClient(CLIENT_TYPE, clientName, metadata)

            if (response.success) {
                saveClientId(username, response.client.clientId)
                return@withContext response.client.clientId
            }

            throw Exception("Nie udało się zarejestrować klienta")
        } catch (e: Exception) {
            throw Exception("Błąd podczas rejestracji klienta: ${e.message}")
        }
    }

    private suspend fun isClientValid(clientId: String, apiClient: ApiClient): Boolean {
        return try {
            val response = apiClient.getClient(clientId)
            response.success
        } catch (e: Exception) {
            false
        }
    }

    private fun getSavedClientId(username: String): String? {
        return try {
            val file = getClientFile(username)
            if (file.exists()) {
                file.readText().trim()
            } else {
                null
            }
        } catch (e: Exception) {
            null
        }
    }

    private fun saveClientId(username: String, clientId: String) {
        try {
            val file = getClientFile(username)
            file.parentFile?.mkdirs()
            file.writeText(clientId)
        } catch (e: Exception) {
            // Ignoruj błędy zapisu
            println("Błąd zapisu ID klienta: ${e.message}")
        }
    }

    private fun removeSavedClientId(username: String) {
        try {
            val file = getClientFile(username)
            if (file.exists()) {
                file.delete()
            }
        } catch (e: Exception) {
            // Ignoruj błędy usuwania
            println("Błąd usuwania ID klienta: ${e.message}")
        }
    }

    private fun getClientFile(username: String): File {
        return File(context.filesDir, "$CLIENT_FILE_PREFIX$username.txt")
    }
}